### PR TITLE
improvements of periodic GW

### PR DESCRIPTION
**Pull request by JWilhelm**
- regularized RI for periodic GW, enables use of large RI basis sets together with the minimum image convention (could be also interesting for periodic RI-Hartree-Fock)
- diagonalization + inversion instead of Cholesky decomposition + inversion
- still to do: k-points for self-energy for computing band structures

### CODE REVIEW

File: src/input_cp2k_mp2.F [hunk 1]
```diff
@@ -734,9 +734,11 @@ SUBROUTINE create_ri_g0w0(section)
       CALL section_add_keyword(section, keyword)
       CALL keyword_release(keyword)
 
-      CALL keyword_create(keyword, __LOCATION__, name="PERIODIC", &
-                          description="If true, the periodic correction scheme is used employing k-points.", &
-                          usage="PERIODIC", &
+      CALL keyword_create(keyword, __LOCATION__, name="PERIODIC_CORRECTION", &
+                          description="If true, the periodic correction scheme is used employing k-points. "// &
+                          "Method is not recommended to use, use instead PERIODIC_LOW_SCALING which much "// &
```
Comment (fstein93) [inline src/input_cp2k_mp2.F:739]: which is much more accurate

File: src/input_cp2k_mp2.F [hunk 2]
```diff
@@ -1126,38 +1134,90 @@ SUBROUTINE create_low_scaling(section)
...
+         "of a slightly worse RI approximation. In case the parameter 0.0 is chosen, ordinary RI is used.", &
+         usage="REGULARIZATION_RI 1.0E-4", &
+         default_r_val=1.0E-2_dp)
+      CALL section_add_keyword(section, keyword)
+      CALL keyword_release(keyword)
+
+      CALL keyword_create( &
+         keyword, __LOCATION__, &
+         name="MAKE_CHI_POS_DEFINITE", &
+         description="If true, makes and eigenvalue decomposition of chi(iw,k) and removes negative "// &
```
Comment (fstein93) [inline src/input_cp2k_mp2.F]: Maybe: makes eigenvalue decomposition of chi(iw,k) positive semi-definite by removing negative values.

File: src/rpa_util.F [hunk 3]
```diff
@@ -934,13 +934,18 @@ SUBROUTINE get_sub_para_kp(fm_struct_sub_kp, para_env, nkp, dimen_RI, &
...
+      ! JW we use all processors for every k-point, a subgroup does not improve the
+      ! performance substantially
+
+!      IF (nkp > para_env%num_pe .OR. do_kpoints_cubic_RPA) THEN
+!         ! we have all kpoints on every processor
+!         num_proc_per_kp = para_env%num_pe
+!      ELSE
+!         ! we have only one kpoint per group
+!         num_proc_per_kp = para_env%num_pe/nkp
+!      END IF
```
Comment (fstein93) [inline src/rpa_util.F]: I would rather remove it for the production code.

File: src/rpa_main.F [hunk 4]
```diff
@@ -691,6 +694,154 @@ SUBROUTINE rpa_ri_compute_en(qs_env, Erpa, mp2_env, BIb_C, BIb_C_gw, BIb_C_bse_i
...
+!
+!      CALL create_kp_and_calc_kp_orbitals(qs_env, qs_env%mp2_env%ri_rpa_im_time%kpoints_Sigma, &
+!                                          "GENERAL", virtual, para_env%num_pe, &
+!                                          kpgeneral=kpgeneral)
+!
+!      CALL create_kp_and_calc_kp_orbitals(qs_env, qs_env%mp2_env%ri_rpa_im_time%kpoints_Sigma_no_xc, &
+!                                          "GENERAL", virtual, para_env%num_pe, &
+!                                          kpgeneral=kpgeneral, with_xc_terms=.FALSE.)
+!
+!      DEALLOCATE(kpgeneral)
```
Comment (fstein93) [inline src/rpa_main.F]: It is an info for you but not useful for production code.

File: src/rpa_util.F [hunk 5]
```diff
@@ -1134,6 +1144,15 @@ SUBROUTINE dealloc_im_time(fm_mo_coeff_occ, fm_mo_coeff_virt, &
       END DO
       DEALLOCATE (t_3c_O_compressed)
 
+      IF (do_kpoints_from_Gamma) THEN
+         IF (ASSOCIATED(kpoints_G)) CALL kpoint_release(kpoints_G)
+         ! JW later for band structures from GW
+         IF (.FALSE.) THEN
+            IF (ASSOCIATED(kpoints_Sigma)) CALL kpoint_release(kpoints_Sigma)
+            IF (ASSOCIATED(kpoints_Sigma_no_xc)) CALL kpoint_release(kpoints_Sigma_no_xc)
+         END IF
```
Comment (fstein93) [inline src/rpa_util.F:1144]: Same as above.

**Comment by fstein93** [comment]
Maybe, I have overseen it, but have you added a regtest for the regularized RI?

**Comment by JWilhelm** [comment]
Thank you for reviewing!

The description of keywords is now improved. The regularized RI is implicitly tested using the test "regtest-gw-cubic/G0W0_kpoints_from_Gamma.inp" using the default regularization parameter 1E-2. I agree that it could be a good idea to have a separate, explicit test, that is now "regtest-gw-cubic/G0W0_kpoints_from_Gamma_RI_regularization_1E-3.inp".

Concerning comments in the code: I understand that for production code, such comments should be avoided. As most of the comments will be anyway gone in the future (after completing band structure calculations from GW), might it be acceptable to leave the three comments (12 lines in total)?

**Comment by fstein93** [comment]
I am in favor of removing these lines because a new CP2K release is planned for the near future.

**Comment by JWilhelm** [comment]
I hope that it is ok now... I will be happy if you could merge the pull request to see whether the other regtester are ok with the code changes. Thank you!

**Pull request merged**